인수 테스트 기반 TDD 2단계 - 경로 조회 기능 PR 요청 드립니다.

ssonsh opened the pull request:
안녕하세요 리뷰어님! 2단계 경로 조회 기능 PR 요청 드립니다. :)

이번 미션을 통해 Outside In, Inside Out 두가지를 경험해보기 위해 접근해보았는데요.
실제로 작업하다보니 인수조건에 의해 인수 테스트를 만들고 난 다음 Outside In으로써 서비스 레이어의 테스트까지는 진행햐지 않고
실패하는 인수테스트를 두고 Inside Out으로 접근하여 기능을 만들어나갔습니다.
이 부분에서 서비스 레이어에 대해  Mock 을 이용한 테스트 코드까지 만들어졌어야 할지 진행하면서도 고민이 있었습니다.

미션의 의도는 둘다 경험을 하는 것을 목표로 하였기 때문에 이 부분에 대해선 테스트 코드를 만들어가며 학습해볼 예정입니다.
이 부분에 있어서 실제 리뷰어님의 실무에선 두가지 모두 활용하시게 되나요? 궁금합니다!

---

PathFinder의 경우 이후 다이잭스트라 알고리즘이 아니라 다른 알고리즘으로도 대체될 수 있을 것이다는 가정을 두고
findShortPath 라는 메소드를 가지는 인터페이스를 구성하고 이를 구현하도록 했는데요 
이 부분에서 구현체를 생성하는 방식을 정적 팩토리 메소드를 이용해 생성하는 방법을 사용하게 되는 경우 static 메소드라 
인터페이스에서 선언해두고 Overriding 할 수 없는데요. 이 부분에서는 객체 생성방법을 어떤 방식을 주로 사용하시나요? 

- 객체를 생성한 다음 그래프를 구성하는 메소드를 선언해두어 이를 구현체에서 Overriding 하여 구현하도록 할 수 있을 것 같다는 생각이 들었습니다.

바쁘신 와중에 리뷰해주셔서 감사드리고 적극적인 피드백 부탁드립니다. :)

ssonsh committed:
refactor (LineService) : Transaction 애노테이션의 readOnly 속성 적용

ssonsh committed:
refactor (Distance) : 미사용 메소드 제거!

ssonsh committed:
docs(README) : 2단계 미션 진행을 위한 기능 목록 1차 정리

ssonsh committed:
test(PathAcceptanceTest) : 최단 경로 조회 인수 테스트 구성

- 출발역에서 도착역 까지 최단거리를 조회하는 인수테스트 작성

ssonsh committed:
refactor : RequestParam 네이밍 변경

ssonsh committed:
test(PathAcceptanceTest) : 최단 거리 조회에 대한 에외 케이스 인수조건 정의 및 인수 테스트 작성

- 출발역과 도착역이 같은 경우
- 노선에 없는 역을 출발역으로 조회하는 경우
- 노선에 없는 역을 도착역으로 조회하는 경우

ssonsh committed:
feat (PathFinder) : 다이잭스트라 알고리즘을 이용한 최단거리 조회 Finder 구현

- findShortPath 메소드를 가지는 PathFinder 인터페이스를 구성하고 이를 구현하도록 함.

- 알고리즘이 바뀌더라도 처리될 수 있는 구조로 고민

ssonsh committed:
test (PathFinder) : 최단 거리 찾기에 대한 예외케이스 테스트 코드 작성

- 존재하지 않는 출발역으로 조회하는 경우 (IllegalArgumentException)
- 존재하지 않는 도착역으로 조회하는 경우 (IllegalArgumentException)
- 출발역과 도착역이 같은경우 Exception은 발생되지 않고 조회된 역이 조회한 역 1개이며 거리는 0으로 결과가 반환된다.

ssonsh committed:
test (PathFinder) : 최단 거리 찾기에 대한 예외케이스 테스트 코드 작성

- 존재하지 않는 출발역으로 조회하는 경우 (IllegalArgumentException)
- 존재하지 않는 도착역으로 조회하는 경우 (IllegalArgumentException)
- 출발역과 도착역이 같은경우 Exception은 발생되지 않고 조회된 역이 조회한 역 1개이며 거리는 0으로 결과가 반환된다.

ssonsh committed:
test (PathAcceptanceTest) : 정의했던 인수조건을 패스하도록 Controller, Service 로직 대응

- 기존 Background 정보를 처리하는 BeforeEach에서 잘못된 부분 수정 (교대역/남부터미널/양재 구간)

- 출발역과 도착역이 같은 경우 실제 경로찾기 로직 상 예외가 발생되지 않음으로 정상 인수 조건으로 변경하여 대응

dhmin5693 commented on src/main/java/nextstep/subway/path/dto/PathResponse.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.List;
+import nextstep.subway.station.dto.StationResponse;
+
+public class PathResponse {
```
Spring에서 ResponseBody를 매핑할 때 기본 생성자가 없으면 오류가 납니다.
이 오류는 intelliJ에서 별도의 설정을 변경하지 않았다면 `AcceptanceTest`에서 거르지 못하고 있어요.
Gradle 옵션에서 `Build and run using`, `Run tests using`을 `IntelliJ IDEA`로 변경해주세요.

ssonsh replied:
실제로 API를 호출해보았으면 알 수 있었던 부분을 단순히 gradle clean test 로만 확인하고 넘어갔던 부분에 대해 깊이 반성하게되네요.
!!

dhmin5693 replied:
다른덴 다 추가되었는데 `PathResponse` 에 기본 생성자가 없어요 😅

dhmin5693 commented:
1단계 미션에서 피드백을 드렸어야 하는 내용인데 신경쓰지 못했네요. 뒷북 죄송합니다 🙇‍♂️
---
혹시 프론트엔드를 실행해보셨을까요? 
아마 제대로 작동이 안될거에요. Response가 기존 스펙이랑 달라졌기 때문입니다.
`LineName` 등은 엔티티의 일부이기 때문에 Response까지 내리시면 안됩니다.
이 부분 수정해주세요!

dhmin5693 commented on src/main/java/nextstep/subway/path/dto/PathResponse.java:
```diff
@@ -0,0 +1,26 @@
+import nextstep.subway.station.dto.StationResponse;
+
+public class PathResponse {
+    private List<StationResponse> stations;
```
미션에서 제시된 Response 스펙과 다릅니다.
또한 `StationResponse` 에서는 엔티티가 포함되어 있습니다.
엔티티는 presentation layer까지 도달하면 안돼요.

지난 번에 피드백을 드렸어야 하는 부분인데 누락이 되었네요.
지금 프론트와 연계하여 테스트를 해보시면 아마 정상 실행이 안될거에요.
`StationResponse` 뿐만이 아니라 `LineResponse` 등 엔티티를 내려준 모든 response가 기존과 스펙이 바뀌었어요.

프론트를 실행해보시고 정상적으로 나오지 않는 부분은 전부 수정해주세요.
확인 결과 프론트 코드는 문제가 없으니 response 스펙만 맞춰주시면 됩니다.

---

expected
```json
{
  "id": 5,
  "name": "양재시민의숲역",
  "createdAt": "2020-05-09T23:54:12.007"
}
```

---

actual
```json
{
  "id": 5,
  "name": {
    "name": "양재시민의숲역"
  },
  "createdAt": "2020-05-09T23:54:12.007"
}
```

ssonsh replied:
넵 :) API Response 가 자연스럽게 변경되어 버렸었네요! 그리고 그걸 확인도 못하고 있었습니다 ㅠㅠ

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -29,10 +29,6 @@ public boolean isLessThanOrEqualsTo(Distance distance) {
         return this.distance <= distance.distance;
     }
 
```
수정한 부분에 대해서만 코멘트를 달 수 있어 리뷰 대상과 다른 곳에 코멘트를 달았어요. 양해부탁드려요.
지난 리뷰 때 미처 확인하지 못한 부분입니다. 🙇‍♂️

`@Embeddable` 클래스 필드에 `@Column`이 명시되어 있지 않아서 예상치 못한 동작을 할 수 있습니다.
https://www.inflearn.com/questions/78106

ssonsh replied:
맞습니다 ㅠㅠ 이 부분도 제대로 챙기지 못하였네요.

dhmin5693 commented:
질문에 대한 답변 드립니다.

---

실무에서는 TDD를 적용하기 쉽지 않습니다. 제일 큰 이유는 시간이겠네요.
같은 피쳐를 개발해도 TDD가 시간이 조금 더 많이 소모됩니다.
또 이미 존재하는 기능을 바꾸는 경우도 많은데 이럴때도 TDD로 개발하기는 참 어렵거든요.
조금 여유로울 때에는 가급적 TDD를 해보려고 노력하고 있어요.

언급하신 내용 중 굳이 방식을 고르자면 저는 Outside In 방식을 선호합니다.
controller에 작성해야 하는 내용은 이미 정해져 있습니다.
사실상 내부 구현은 service layer부터 시작이 될텐데요,
service에서는 response를 내려주어야 하기 때문에 사실상 반환 유형은 고정이 됩니다.
그렇게 위부터 차례대로 작성하는 Top down 방식이 생각을 순서대로 연결하기에 더 편하더라구요.

---

인터페이스는 객체의 동작을 명시하는 일종의 설계도입니다. 
인터페이스가 생성하는 부분까지 알 필요가 있을까요?
생성의 역할은 각자 클래스에게 맡겨두는게 더 적절합니다.

다익스트라 외의 알고리즘으로도 구현할 수 있다는 접근 방식은 매우 좋습니다.
만약 에이스타 알고리즘 라이브러리를 사용하면 어떨까요?
현재 라이브러리는 필드에 넣어두고 사용할 수 있어서 생성자가 필요하지만, 다른 라이브러리는 활용 방식부터가 달라 생성자에 주입하는 데이터가 필요없을 수도 있습니다.

`PathFinder`라는 이름에 맞게 findShortPath라는 행위만 보장해도 충분합니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/finder/DijkstraPathFinder.java:
```diff
@@ -0,0 +1,57 @@
+package nextstep.subway.path.finder;
```
finder라는 분류는 다소 어색하다고 느껴집니다.
도메인 로직에 가까운 부분임을 고려하여 적절하게 옮겨보는건 어떨까요?

ssonsh replied:
path 아래 domain의 명칭으로 가져가는게 더 좋을까요? :)

ssonsh replied:
피드백 대응의 작업은 path.domain 패키지 하위에 Path 라는 도메인 객체를 구성하였고 PathFinder를 위치시켰습니다! :)

dhmin5693 replied:
적절하네요 👍

dhmin5693 commented on src/main/java/nextstep/subway/path/finder/PathFinder.java:
```diff
@@ -0,0 +1,8 @@
+import nextstep.subway.station.domain.Station;
+
+public interface PathFinder {
+    PathResponse findShortPath(Station sourceStation, Station targetStation);
```
가능성을 염두한 설계 💯 

`PathFinder`에서 `PathResponse` 를 반환하면 가능성을 고려하신 의도가 조금 흐려질 수 있다는 생각이 듭니다.
response를 반환하니 사실상 `PathService`에서만 사용할 수 있을 것처럼 느껴지거든요.

다른 도메인 클래스를 반환하고 서비스 단에서 response로 변환하는건 어떤가요?
그러면 `PathFinder`의 활용성이 더 나아질거에요.

ssonsh replied:
아! 네 맞습니다 Response를 반환하다보니 Service 외에서 사용한다면 어색하다고 느낄 수 있다고 생각됩니다.
이 부분에 대해서는 반환을 Path 라는 클래스로 반환하고 이 Path 를 이용해 Response를 만들어내는 방향으로 개선해보도록 하겠습니다.
- 이 부분에서 Path 라는 네이밍이 적절한지도 고민스럽긴 합니다! ㅎ

dhmin5693 replied:
`Path` 도 괜찮다고 생각해요. `ShortestPath` 가 조금 더 의미를 드러낼 순 있겠네요.

dhmin5693 commented on src/main/java/nextstep/subway/utils/StreamUtils.java:
```diff
@@ -6,19 +6,30 @@
         return collections.stream().filter(predicate).findFirst();
     }
+
+    public static <T, R> List<R> flatMapToList(Collection<T> collections,
```
개인적인 의견이지만 `StreamUtils`가 진짜 필요한지 조금 고민이 됩니다.
일부 로직을 묶어 메소드로 만드는 이유는 코드를 재사용할 수 있게 묶어주고 이름으로 가독성까지 챙기는 잇점을 가져갈 수 있기 때문입니다.

`Stream` API의 장점은 무엇일까요?
저는 체이닝 방식에서 오는 가독성이 크다고 생각해요.
순서대로만 읽으면 로직을 충분히 해석할 수 있게 설계었으니까요.

위의 두 클래스까지는 메소드명에서 어느 정도 유추할 수 있겠지만, 이 메소드에서 실행하는 스트림은 동작이 너무 많아 메소드명으로 표현하기 어렵습니다.
게다가 매개변수는 3개 중 2개가 `FunctionalInterface`라서 읽기가 쉽지 않거든요.
오히려 가독성에 방해가 된다고 봅니다.

ssonsh replied:
네 저도 공감하는 부분입니다. :)
"Stream API의 장점인 체이닝 방식에서 오는 가독성" 이 부분에 대해 공감하고 그렇기 때문에 현재 구성되어 있는 StreamUtil이 그렇게 까지 활용도가 높을까? 라는 의문이 드는 것도 자연스러운 것 같습니다! 

이 부분에 대해서도 제가 근무하고 있는 곳에서 나름의 컨벤션으로 가져가고 있어서 자연스럽게 작업하지 않았나 싶습니다. :)

익숙하지 않은 사람이 보았을 때 가독성이 낮다면 그 또한 변경이 필요할 수 있는 컨벤션이라고 생각되네요! :)

좋은 의견 감사드립니다!

ssonsh replied:
오히려 체이닝과정에서 복잡한 코드가 생성되면 그 부분을 가독성 좋게 메소드 추출하는 것이 더 좋은 방향이라고 개인적으로 생각되네요!

dhmin5693 reviewed (changes requested):
미션 진행하시느라 수고하셨습니다.

꼭 확인을 해주셔야 하는 코멘트, 그리고 리뷰이님의 생각을 듣고자 하는 코멘트를 남겼습니다.
확인 후 다시 요청 부탁드릴게요 🙏

ssonsh committed:
refactor : Embeddable 클래스에 대해 Column 애노테이션 명시

ssonsh committed:
refactor : Response 모델에서 사용된 값객체 제거 & 기본 생성자 추가

ssonsh committed:
refactor : Response 모델에서 사용된 값객체 제거 & 기본 생성자 추가

- 변경에 따른 테스트 코드 실패 대응

- 값 객체가 없는 상황에서 내부 값을 조회하려고 할 때 NPE 발생

ssonsh committed:
refactor : PathFinder 의 위치를 적절한 위치로 이동

- asis : path.finder
- tobe : path.domain

ssonsh committed:
refactor : StreamUtils 에 대한 고찰

- 활용하는 부분에서 익숙하지 않다면 더 혼란을 가중시킬 수 있는 부분임에 공감

- 체이닝으로 충분히 가독성을 가져갈 수 있다고 생각됨.

- 만약 체이닝되는 부분의 코드가 복잡해지면 그 부분을 메소드로 추출하는게 좋다고 생각됨.

ssonsh committed:
refactor : PathFinder의 반환을 Response로 고정하지 않는다.

- 추가적인 반환 객체를 만들며 이를 통해 Response를 만들어내거나

- 부수적인 작업을 할 수 있또록 한다.

ssonsh committed:
refactor (Line Name, Color) : 값 객체에서 의도한 것과 다른 방향임으로 변경

ssonsh committed:
refactor (Path) : 경로 도메인에 대한 네이밍 변경

- asis : ShortPath
- tobe : Path

- ShortPath는 경로 중 짧은 경로의 개념이므로 오히려 Path 라는 네이밍으로 가져가는게 더 적절하다고 판단되었습니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,33 @@
+        return new Path(stations, distance);
+    }
+
+    public PathResponse convertPathResponse() {
```
마지막으로 하나만 더 남겨보자면, `PathResponse` 는 presentation 영역의 클래스고 `Path`는 domain 영역의 클래스입니다.
domain이 presentation을 참조하는 형태이지요.

각 영역을 원으로 표기한다면 persistence 영역이 가장 가운데, presentation이 가장 바깥의 원으로 표현될 것입니다.
presentation이 바깥인 이유는 외부와 통신하는 영역이기 때문입니다.
바깥 영역의 객체는 안쪽 영역을 참조할 수 있지만 반대로는 불가능합니다.

코드가 한 방향으로 흐르지 않으면 다소 난잡해질 수 있으니 영역간 침범에 주의해주세요.

ssonsh replied:
domain 영역에 있는 Path에서 Presentation 영역의 PathResponse를 처리하고 있네요 ㅠㅠ 
바깥 영역의 PathResponse가 안쪽 영역의 Path를 참조하여 처리하는 방향이 정방향이라 할 수 있고,
그 반대로 제가 구현해놓은 것은 Path 안쪽 영역이 PathResponse 바깥쪽 영역을 참조하여 처리하게 됨으로써
역방향이 되어 버렸네요 

이 부분에 대해서 충분히 인식하며 작업할 수 있도록 해야 할 것 같습니다.! 

해당 내용은
PathResponse가 Path를 이용하여 원하는 객체를 생성하는 방향으로 리팩토링 하도록 하겠습니다. :)

추후 Web Layer와 Domain Layer가 모듈로 분리가 된다고 하였을 때 바로 이슈가 발생될 부분이기도 한 것 같습니다. :)

dhmin5693 reviewed (changes requested):
아침부터 과제하느라 고생 많으셨어요.
딱 하나 반영 안된게 있어서 다시 확인 부탁드리겠습니다.
그리고 피드백을 반영해주셔서 좋은 형태로 만들어주신 부분이 있는데 여기도 하나 남겼으니 같이 확인 부탁드릴게요 🙇‍♂️

ssonsh committed:
refactor (Path, PathResponse) : Path와 PathResponse 간의 참조 방향이 잘못 정의 되어 있음.

- asis : Path 도메인 영역에서 PathResponse 프리젠테이션 영역으로 참조를 하고 있음.

- tobe : PathResponse 프리젠테이션 영역에서 Path 도메인 영역을 참조하도록 변경

ssonsh committed:
refactor (PathResponse) : PathResponse의 기본생성자가 누락된 부분 추가!

dhmin5693 reviewed (approved):
피드백 반영하시느라 수고 많으셨습니다.
다음 미션 진행해주세요 😄
